pkg/image/admission/admission.go

```diff
@@ -0,0 +1,302 @@
+
+// NewImageLimitRangerPlugin provides a new imageLimitRangerPlugin.  Also see SetOpenshiftClient.
+func NewImageLimitRangerPlugin(client clientset.Interface, config io.Reader) (*imageLimitRangerPlugin, error) {
+	plugin := &imageLimitRangerPlugin{
+		Handler:          kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		imageStreamStore: cache.NewStore(cache.MetaNamespaceKeyFunc),
+	}
+	limitRanger, err := limitranger.NewLimitRanger(client, plugin)
+	if err != nil {
+		return nil, err
+	}
+	plugin.limitRanger = limitRanger
+
+	// NOTE: the reflector is set up in SetOpenshiftClient to ensure we have access to the client
+	return plugin, nil
+}
+
+// Run starts the reflectors used by the plugin.
+func (a *imageLimitRangerPlugin) Run() {
+	if a.stopChan == nil {
+		a.stopChan = make(chan struct{})
+		a.imageStreamReflector.RunUntil(a.stopChan)
+	}
+}
+
+// Stop stops the reflectors used by the plugin.
+func (a *imageLimitRangerPlugin) Stop() {
+	if a.stopChan != nil {
+		close(a.stopChan)
+		a.stopChan = nil
+	}
+}
+
+// Admit invokes the admission logic for checking against LimitRanges.
+func (a *imageLimitRangerPlugin) Admit(attr kadmission.Attributes) error {
+	if !a.SupportsAttributes(attr) {
+		return nil // not applicable
+	}
+
+	// If the namespace is not set on the object ensure it gets set.
+	if om, err := kapi.ObjectMetaFor(attr.GetObject()); err == nil {
+		if om.Namespace == "" {
+			glog.V(5).Infof("%s: namespace was not set for object, defaulting to %s", PluginName, attr.GetNamespace())
+			om.Namespace = attr.GetNamespace()
+		}
+	}
+
+	return a.limitRanger.Admit(attr)
+}
+
+// SupportsAttributes is a helper that returns true if the resource is supported by the plugin.
+// Implements the LimitRangerActions interface.
+func (a *imageLimitRangerPlugin) SupportsAttributes(attr kadmission.Attributes) bool {
+	if attr.GetSubresource() != "" {
+		return false
+	}
+
+	resource := attr.GetResource()
+	return resource == imageapi.Resource("imagestreammappings") ||
+		resource == imageapi.Resource("imagestreamtags") ||
+		resource == imageapi.Resource("imagestreams")
+}
+
+// SupportsLimit provides a check to see if the limitRange is applicable to image objects.
+// Implements the LimitRangerActions interface.
+func (a *imageLimitRangerPlugin) SupportsLimit(limitRange *kapi.LimitRange) bool {
+	if limitRange != nil {
+		for _, limit := range limitRange.Spec.Limits {
+			if limit.Type == imageapi.LimitTypeImageSize {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// Limit is the limit range implementation that checks resource against the
+// image limit ranges.
+// Implements the LimitRangerActions interface
+func (a *imageLimitRangerPlugin) Limit(limitRange *kapi.LimitRange, kind string, obj runtime.Object) error {
+	var images map[string]imageapi.Image
+
+	switch isObj := obj.(type) {
+	case *imageapi.ImageStreamMapping:
+		images = map[string]imageapi.Image{isObj.Image.Name: isObj.Image}
```

Comment:
If it's a performance optimization, let's treat it like one.  I think this
invalidates the concern with quota not being enforced in the registry (or
drops its importance) - the manifest creation is the authoritative check,
not the layer.  It can then use a cache and periodically refresh the values
(limits are valid for X minutes).

As far as pruning pushed layers, we have to eventually add that, otherwise
someone can break the cluster.  Isn't there the upload cache cleaner that
goes and removes layers that aren't completely matched?

On Wed, Mar 23, 2016 at 3:56 AM, Michal Minar <notifications@github.com>
wrote:

> In pkg/image/admission/admission.go
> <https://github.com/openshift/origin/pull/7989#discussion_r57120608>:
>
> > +				return true
> > +			}
> > +		}
> > +	}
> > +	return false
> > +}
> > +
> > +// Limit is the limit range implementation that checks resource against the
> > +// image limit ranges.
> > +// Implements the LimitRangerActions interface
> > +func (a *imageLimitRangerPlugin) Limit(limitRange *kapi.LimitRange, kind string, obj runtime.Object) error {
> > +	var images map[string]imageapi.Image
> > +
> > +	switch isObj := obj.(type) {
> > +	case *imageapi.ImageStreamMapping:
> > +		images = map[string]imageapi.Image{isObj.Image.Name: isObj.Image}
>
> It's just a performance optimalization. The admission check here is
> triggered a bit late - once all the layers are already pushed. With an
> additional check in the registry, we can safe a minute of useless blob
> pushing and storage needlessly occupied (which is a big issue until we
> refactor image pruning). Right now there's no code in origin/registry that
> will prune pushed layers whose manifest is denied.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7989/files/5c9a396909ea2fbcde785ee540ca1311461a2e9e#r57120608>
>